writeaheadlog_integration_test.go

```diff
@@ -1,556 +1,487 @@
-
-				// Add a broken count. Because the break is after the WAL commits, the
-				// count should still restore correctly on the next iteration where we
-				// call 'newCountdown'.
-				err = cd.addCountBroken()
-				if err != nil {
-					t.Fatal(err)
-				}
-				expectedCountdownLen++
-				err = cd.Close()
-				if err == nil {
-					t.Fatal("Should have returned an error but didn't")
-				}
+		// Reset dependencies and increase the write limit to allow for higher
+		// success rate
+		deps.reset()
+		*deps.writeLimit = 5000
+
+		var siloOff int64
+		var siloOffsets []int64
+		var silos = make(map[int64]*silo)
+
+		// Start the threaded update after creating and initializing the silos
+		for i := 0; int64(i) < numSilos; i++ {
+			wg.Add(1)
+			silo := newSilo(siloOff, 1+i*numIncrease, deps, file)
+			if err := silo.init(); err != nil {
+				t.Fatalf("Failed to init silo: %v", err)
 			}
 
-			// Check at this point that the wal is less than 100 pages.
-			info, err := os.Stat(filepath.Join(dir, "wal.dat"))
-			if err != nil {
-				t.Fatal(err)
-			}
-			if info.Size() > 100*pageSize {
-				t.Error("the wal is too large")
-			}
+			go silo.threadedUpdate(t, wal, testdir, &wg)
+			siloOffsets = append(siloOffsets, siloOff)
+			silos[siloOff] = silo
+			siloOff += int64(len(silo.numbers)*4) + checksumSize
 		}
 
-		// Test the parallelism. Basic way to do that is to have a second file that
-		// we update in parallel transactions. But I'd also like to be able to test
-		// parallel transactions that act on the same file? Not sure if that's
-		// strictly necessary. But we could have a second file that perhaps tracks
-		// two unrelated fields, like 5 integer arrays that are all intialized with
-		// the same integers, over 10kb or something. And then that file could have
-		// 3 independent sets of these things, so they all have clear dependence
-		// within but no dependence next to. Then we'll update all of them and the
-		// count as well in parallel transactions.
-		if r&(1<<3) == 0 {
-			cd, err := newCountdown(dir)
-			if err != nil {
-				t.Fatal(err)
-			}
-			var wg sync.WaitGroup
-			for i := uint64(0); i < 50; i++ {
-				wg.Add(1)
-				go func(i uint64) {
-					defer wg.Done()
-					for j := uint64(1); j < 200; j++ {
-						err := cd.changeSplotch(i, j)
-						if err != nil {
-							t.Error(err)
-						}
-					}
-				}(i)
-			}
-			wg.Wait()
-			err = cd.Close()
-			if err != nil {
-				t.Fatal(err)
-			}
-		}
+		// Corrupt the disk
+		deps.disable(false)
 
-		// Open and close the cd to allow the consistency checks to run.
-		cd, err := newCountdown(dir)
-		if err != nil {
-			t.Fatal(err)
+		// Wait for all the threads to fail
+		wg.Wait()
+
+		// Close wal
+		if err := wal.logFile.Close(); err != nil {
+			t.Error(err)
 		}
-		err = cd.Close()
+
+		// Reset dependencies and set limit to a high value that can never be
+		// reached to make sure we can recover after enough tries
+		deps.reset()
+		*deps.writeLimit = math.MaxUint64
```

Comment:
No need to set the write limit that high, you should be able to recover with a write limit of 25,000 or whatever.

Reply:
I set it that high so I can be 100% sure it is never reached. When 250 transactions with somewhere around 500 updates on average are recovered 10 times we hit limits > 100.000